generateReusableSchemas (zod): recursive schema references hoisted sub-models (inline enum/object) it never emits → TS2304

### Description

With `client: "zod"` + `override.zod.generateReusableSchemas: true`, a **recursive** component schema (one that sits in a `$ref` cycle, so it's pinned to an explicit `const X: zod.ZodType<X>`) generates a hand-written TS type body that references the implicit sub-models orval mints for that schema's **inline enums and inline nested objects** — but those sub-model declarations are never emitted. The output references a name that doesn't exist.

The trigger is specifically the recursive path. A non-recursive schema derives its public type as `zod.input<typeof X>`, which never names the hoisted sub-models, so it's unaffected. Only schemas forced onto the `zod.ZodType<X>` annotation (because they're in a cycle) hit this.

### Output client

zod

### Configuration (orval.config)

```ts
import { defineConfig } from 'orval';

export default defineConfig({
  repro: {
    input: { target: './openapi.yaml' },
    output: {
      mode: 'single',
      client: 'zod',
      target: './out/repro.zod.ts',
      override: {
        zod: { generateReusableSchemas: true },
      },
    },
  },
});
```

### Environment

- orval: 8.19.0 (also reproduces on 8.16.0)
- @orval/zod: 8.19.0
- zod: 3.x
- node: 22.x

### Expected behavior

`ActionMeta` (the implicit sub-model orval hoists for the inline `meta` object) should be emitted alongside `Action`, the same way the TypeScript model client emits hoisted sub-models, so the generated type compiles.

### Actual behavior

The generated file references `ActionMeta` but never declares it, so it doesn't compile:

```ts
export type Action = {
  next?: Action;
  meta?: ActionMeta;   // <- never declared
};

export const Action: zod.ZodType<Action> = zod.object({
  "next": zod.lazy(() => Action).optional(),
  "meta": zod.object({
  "label": zod.string().optional()
}).optional()
});
```

```
out/repro.zod.ts(11,10): error TS2304: Cannot find name 'ActionMeta'.
```

(In `mode: 'split'` the same cause surfaces as `TS2305: Module './schemas' has no exported member 'ActionMeta'`, and an inline `enum` property hoists the same way — e.g. `ActionKind` — so a real-world schema with a recursive variant carrying both an inline enum discriminator and a nested object emits several of these per variant.)

### OpenAPI document (minimal, if applicable)

```yaml
openapi: 3.1.0
info:
  title: Recursive reusable-schema sub-model repro
  version: 1.0.0
paths:
  /action:
    get:
      operationId: getAction
      responses:
        '200':
          description: ok
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Action'
components:
  schemas:
    Action:
      type: object
      properties:
        next:
          $ref: '#/components/schemas/Action'   # self-cycle -> recursive -> explicit zod.ZodType<Action>
        meta:                                    # inline object -> hoisted to ActionMeta, never emitted
          type: object
          properties:
            label:
              type: string
```

### Additional context

The hoisted sub-models are produced by `resolveValue` and returned in its `.schemas` array. The recursive branch of `renderReusableSchemaEntry` (`packages/orval/src/write-zod-specs.ts`) consumes `.value` and `.imports` but drops `.schemas`, so the type body references names that were never declared.

--- a/packages/orval/src/write-zod-specs.test.ts
+++ b/packages/orval/src/write-zod-specs.test.ts
@@ -616,6 +616,70 @@ describe('writeZodSchemas with generateReusableSchemas', () => {
 
     await fs.remove(root);
   });
+
+  it('emits the implicit sub-model an inline nested object hoists in a recursive schema', async () => {
+    // A recursive schema (self-loop via `next`) takes the explicit
+    // `zod.ZodType<T>` path, whose TS body is hand-written from
+    // `resolveValue().value`. That body references the implicit sub-model
+    // `resolveValue` mints for the inline `meta` object (`ActionMeta`), which
+    // arrives in `resolved.schemas`. Pre-fix the writer dropped
+    // `resolved.schemas`, so the body named a type that was never declared
+    // (TS2552 in single-file output, TS2305 in split). The acyclic path never
+    // hits this: it derives its type via `zod.input<typeof X>`.
+    const root = await fs.mkdtemp(path.join(tmpdir(), 'orval-zod-reuse-sub-'));
+    const schemasPath = path.join(root, 'schemas');
+
+    const builder = {
+      spec: {
+        components: {
+          schemas: {
+            Action: {
+              type: 'object',
+              properties: {
+                next: { $ref: '#/components/schemas/Action' },
+                meta: {
+                  type: 'object',
+                  properties: { label: { type: 'string' } },
+                },
+              },
+            },
+          },
+        },
+      },
+      target: '',
+      schemas: [
+        { name: 'Action', schema: { $ref: '#/components/schemas/Action' } },
+      ],
+    } satisfies Parameters<typeof writeZodSchemas>[0];
+
+    const options = createOutputOptions();
+    options.override.zod.generateReusableSchemas = true;
+
+    await writeZodSchemas(builder, schemasPath, '.ts', '', options);
+
+    const actionContent = await fs.readFile(
+      path.join(schemasPath, 'Action.ts'),
+      'utf8',
+    );
+
+    // The recursive path is taken (so the hand-written body is in play) and
+    // references the hoisted sub-model.
+    expect(actionContent).toContain(
+      'export const Action: zod.ZodType<Action> = ',
+    );
+    expect(actionContent).toMatch(/meta\??: ActionMeta/);
+    // The sub-model is declared locally in the same file — not a dangling
+    // reference, not a cross-file import, and no sibling file is written for
+    // it (it isn't a component).
+    expect(actionContent).toContain('export type ActionMeta = ');
+    expect(actionContent).not.toMatch(/from '\.\/ActionMeta'/);
+    expect(await fs.pathExists(path.join(schemasPath, 'ActionMeta.ts'))).toBe(
+      false,
+    );
+    expect(actionContent).not.toContain('__REF_');
+
+    await fs.remove(root);
+  });
 });
 
 describe('buildSiblingImports', () => {

--- a/packages/orval/src/write-zod-specs.ts
+++ b/packages/orval/src/write-zod-specs.ts
@@ -311,7 +311,16 @@ function renderReusableSchemaEntry(
     // emitted here or the body references undeclared names (TS2552). The
     // acyclic branch never hits this — it derives its type via
     // `zod.input<typeof X>`, so it never names them.
-    const subModels = resolved?.schemas ?? [];
+    // Dedupe by name: `resolveValue` can surface the same sub-model twice
+    // (e.g. an allOf/oneOf branch re-resolving a shared inline shape), and two
+    // `export type X` for one name is a TS2300 duplicate-identifier error. The
+    // model writer dedupes the same way (group-by-name in writers/schemas.ts).
+    const seenSubModels = new Set<string>();
+    const subModels = (resolved?.schemas ?? []).filter((s) => {
+      if (seenSubModels.has(s.name)) return false;
+      seenSubModels.add(s.name);
+      return true;
+    });
     const subModelBlock = subModels.length
       ? `${subModels.map((s) => s.model.trimEnd()).join('\n')}\n\n`
       : '';

--- a/packages/orval/src/write-zod-specs.ts
+++ b/packages/orval/src/write-zod-specs.ts
@@ -304,6 +304,19 @@ function renderReusableSchemaEntry(
       ? resolveValue({ schema, name: entry.name, context })
       : undefined;
     const typeBody = resolved ? resolved.value : 'unknown';
+    // The recursive type body is hand-written from `resolved.value`, which
+    // references the implicit sub-models `resolveValue` generates for inline
+    // enums and nested objects (e.g. `<Name>Type`, `<Name>Target`,
+    // `<Name><NestedProp>`). Those live in `resolved.schemas` and MUST be
+    // emitted here or the body references undeclared names (TS2552). The
+    // acyclic branch never hits this — it derives its type via
+    // `zod.input<typeof X>`, so it never names them.
+    const subModels = resolved?.schemas ?? [];
+    const subModelBlock = subModels.length
+      ? `${subModels.map((s) => s.model.trimEnd()).join('\n')}\n\n`
+      : '';
+    // Sub-models are declared locally above, so they're never imports.
+    const localNames = new Set(subModels.map((s) => s.name));
     // Dedupe by local binding name (`alias ?? name`). When `resolveValue`
     // surfaces the same component twice — e.g. once aliased, once not —
     // collapsing on the binding key keeps both the file's import list and the
@@ -311,7 +324,9 @@ function renderReusableSchemaEntry(
     const seen = new Set<string>();
     const extraImports: ExtraImport[] = [];
     for (const imp of resolved?.imports ?? []) {
-      if (!imp.name || imp.name === entry.name) continue;
+      if (!imp.name || imp.name === entry.name || localNames.has(imp.name)) {
+        continue;
+      }
       const bindingKey = imp.alias ?? imp.name;
       if (seen.has(bindingKey)) continue;
       seen.add(bindingKey);
@@ -323,7 +338,7 @@ function renderReusableSchemaEntry(
 
     return {
       content:
-        `${consts}export type ${entry.name} = ${typeBody};\n\n` +
+        `${consts}${subModelBlock}export type ${entry.name} = ${typeBody};\n\n` +
         `export const ${entry.name}: zod.ZodType<${entry.name}> = ${entry.zod};\n\n` +
         `export type ${entry.name}Output = zod.output<typeof ${entry.name}>;`,
       extraImports,

